_drafts/2023-04-15-functional-error-handling-in-kotlin.md

```diff
@@ -0,0 +1,811 @@
+
+When we work with nullable types, the compiler forces us to handle the case when the value is `null`. For example, let's change our primary example, handling errors using nullable types. First, we redefined the `Jobs` interface and its implementation to use nullable types:
+
+```kotlin
+interface Jobs {
+    fun findById(id: JobId): Job?
+}
+
+class LiveJobs : Jobs {
+    override fun findById(id: JobId): Job? = try {
+        JOBS_DATABASE[id]
+    } catch (e: Exception) {
+        null
+    }
+}
+```
+
+As we said, using nullable types to handle failures means completely losing the cause of the error, which is not propagated in any way to the caller.
+
+Then, we also change the `JobsService`, and we try to use the dereference operator `.` to access the `salary` property of the `Job` object directly:
+
+```kotlin
+class JobsService(private val jobs: Jobs) {
+    fun retrieveSalary(id: JobId): Double =
+        jobs.findById(id).salary.value
+}
+```
+
+If we try to compile the code, we get the following error:
+
+```text
+Compilation failure
+[ERROR] /functional-error-handling-in-kotlin/src/main/kotlin/in/rcard/nullable/NullableTypeErrorHandling.kt:[21,26] Only safe (?.) or non-null asserted (!!.) calls are allowed on a nullable receiver of type Job?
+```
+
+The compiler is warning us that we forgot to handle the case where the list is `null`. As we can see, the compiler is helping us to avoid errors. Let's fix the code:
+
+```kotlin
+fun retrieveSalary(id: JobId): Double =
+    jobs.findById(id)?.salary?.value ?: 0.0
+```
+
+Here, we used the `?.` operator to call a method on a nullable object only if it's not `null`. We must use the `?.` operator for every method call if we have a chain of calls. Finally, we use the "Elvis" operator, `?:`, as a fallback value, in case the job is `null`.
+
+At first, using a nullable value seems less composable than using, for example, the Java `Optional` type. This last type has a lot of functions, `map`, `flatMap`, or `filter`, which make it easy to compose and chain operations.
+
+Kotlin nullable types have nothing to envy to the Java `Optional` type. In fact, the Kotlin standard library provides a lot of functions to handle nullable types. For example, **the `Optional.map` function is equivalent to the `let` scoping function**:
+
+```kotlin
+// Kotlin SDK
+public inline fun <T, R> T.let(block: (T) -> R): R {
+    contract {
+        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+    }
+    return block(this)
+}
+```
+
+The `let` function takes a lambda function as input, applying it to the receiver type, and returning the result, as the `map` function (ed., [contracts](https://kotlinlang.org/docs/whatsnew13.html#contracts) is a powerful tool of the language to give the compiler some hints).
```

Comment:
This contract thing is pretty cool and powerful - maybe we can write about it some more in another article (or in a course 😛 )